test/common/network/listener_impl_test.cc

```diff
@@ -77,13 +77,19 @@ class ListenerImplTest : public TestBaseWithParam<Address::IpVersion> {
       : version_(GetParam()),
         alt_address_(Network::Test::findOrCheckFreePort(
             Network::Test::getCanonicalLoopbackAddress(version_), Address::SocketType::Stream)),
-        api_(Api::createApiForTest(stats_store_)), dispatcher_(*api_) {}
+        api_(Api::createApiForTest(stats_store_)), dispatcher_(api_->allocateDispatcher()) {}
+
+  Event::DispatcherImpl& dispatcherImpl() {
+    Event::DispatcherImpl* impl = dynamic_cast<Event::DispatcherImpl*>(dispatcher_.get());
```

Comment:
This is the ugliness you mentioned in the issue, right? Can you briefly explain your thought process here so I can reconcile it against the diff?

Reply:
Yup. Actually I have this in 2 places in the PR and I factored out a common helper test-base class, and put the explanation there:
```
  Event::DispatcherImpl& dispatcherImpl() {
    // We need access to the concrete impl type in order to instantiate a
    // Test[Udp]Listener, which instantiates a [Udp]ListenerImpl, which requires
    // a DispatcherImpl to access DispatcherImpl::base_, which is not part of
    // the Dispatcher API.
    Event::DispatcherImpl* impl = dynamic_cast<Event::DispatcherImpl*>(dispatcher_.get());
    RELEASE_ASSERT(impl, "dispatcher dynamic-cast to DispatcherImpl failed");
    return *impl;
  }
```